### ISSUE TITLE
qtox kills 6 out 10 routers.

### ISSUE THREAD

**Issue by laietana** [TokTok/c-toxcore#195]
Hi we did a test with 10 routers and qtox ( windows ) killed 6 out of 10 routers . We had to  disable udp to make it work, which brings the questions , is the implementation of hole-punching correct ? How do other programs holepunch without killing the (same) routers. Why do we need udp holepunching when it could first try to upnp ? The router list includes tplink netgear ZTE  thomson and others ( common routers where I live).

At least you could default udp off for now since I cant even recommend it to somone without have to be like " Oh and you need to go uncheck that udp thing... "

Many thanks,
Lait

**Comment by GrayHatter** [comment]
> At least you could default udp off for now 

That's a client setting. You should also open this issue with your client of choice.

**Comment by laietana** [comment]
True, the default setting is up to the qtox program. But the UDP that kills the routers is in the toxcore.

**Comment by iphydf** [comment]
There is preliminary uPNP support in the irungentoo/new branch. Perhaps it would make sense to make a pull request with those patches.

**Comment by fcore117** [comment]
laietana: Yes our country main cable ISP uses Thompson routers too for internet and for connecting IPTV box behind router. Seems like massive connection count especially in UDP mode causes routers to go mad and freeze.

**Comment by GrayHatter** [comment]
@fcore117 are you talking about internet routers? Or home based wifi and/or
internet sharing devices (commonly called routers, even though 'NAT Hub' is
often a much better name)?

**Comment by fcore117** [comment]
GrayHatter: thomson tg784, Thomson ST780WL, Thomson tg789vn, Inteno DG301A, zte models and similar are most used of adsl routers around here.

**Comment by iphydf** [comment]
Who wants to make a PR cherry-picking the uPNP code from https://github.com/irungentoo/toxcore/commits/new? Look for "Commits on Mar 21, 2016", there are a few related commits. Perhaps @Ansa89 can do it?

**Comment by GrayHatter** [comment]
@iphydf I was talking with @cleverca22 on IRC, and IIRC the conclusion we came to was that UPnP would certainly help with this, it wouldn't solve the issue by itself.

**Comment by fcore117** [comment]
i agree upnp is different topic as it wont help with connection count but helps to create direct connection easily if router has upnp support enabled.

**Comment by iphydf** [comment]
We now have an option: `hole_punching_enabled`. You can set it to `false` by calling `tox_options_set_hole_punching_enabled(options, false);`. Does that help?

**Comment by zetok** [comment]
Affects qTox users: https://github.com/qTox/qTox/issues/4249 , disabling UDP helps.
